Add A Pass to Fold Tensors With a Uniform Value, match sdpa on a few models

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #103330
* __->__ #103600


Adds a Constant Folding pass to the joint graph only targeting tensors which can be replaced with a single value, and then removes no-ops from the graph. This allows us to match sdpa in BertForMaskedLM, AlbertForMaskedLM, and LayoutLMForMaskedLM.

BertForMaskedLM 
Perf: 1.6853 -> 1.933, Memory: 0.9462 -> 1.41

AlbertForMaskedLM
Perf: 1.6620 -> 1.761, Memory: 1.257 -> 1.94

LayoutLMForMaskedLM
Perf: (non cudagraphs) 1.6991 -> 1.939x, Memory: 0.9624 -> 1.50

MobileBertForMaskedLM
Perf: 1.864x -> 1.941x, Memory: 0.94 -> 1.03



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78

eellison committed:
Add A Pass to Fold Tensors With a Uniform Value, match sdpa on MaskedLM

[ghstack-poisoned]

eellison committed:
Update on "Add A Pass to Fold Tensors With a Uniform Value, match sdpa on a few models"



Adds a Constant Folding pass to the joint only targeting tensors which can be replaced with a single value, and then removes no-ops from the graph. This allows us to match sdpa in BertForMaskedLM, AlbertForMaskedLM, and LayoutLMForMaskedLM.

BertForMaskedLM 
Perf: 1.6853 -> 1.933, Memory: 0.9462 -> 1.41

AlbertForMaskedLM
Perf: 1.6620 -> 1.761, Memory: 1.257 -> 1.94

LayoutLMForMaskedLM
Perf: (non cudagraphs) 1.6991 -> 1.939x, Memory: 0.9624 -> 1.50

MobileBertForMaskedLM
Perf: 1.864x -> 1.941x, Memory: 0.94 -> 1.03



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

eellison committed:
Update on "Add A Pass to Fold Tensors With a Uniform Value, match sdpa on a few models"



Adds a Constant Folding pass to the joint graph only targeting tensors which can be replaced with a single value, and then removes no-ops from the graph. This allows us to match sdpa in BertForMaskedLM, AlbertForMaskedLM, and LayoutLMForMaskedLM.

BertForMaskedLM 
Perf: 1.6853 -> 1.933, Memory: 0.9462 -> 1.41

AlbertForMaskedLM
Perf: 1.6620 -> 1.761, Memory: 1.257 -> 1.94

LayoutLMForMaskedLM
Perf: (non cudagraphs) 1.6991 -> 1.939x, Memory: 0.9624 -> 1.50

MobileBertForMaskedLM
Perf: 1.864x -> 1.941x, Memory: 0.94 -> 1.03



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

desertfire commented on test/inductor/test_torchinductor_opinfo.py:
```diff
@@ -520,6 +521,7 @@ class TestInductorOpInfo(TestCase):
     )  # inductor kernels failing this test intermittently
     @skipCUDAIf(not HAS_CUDA, "Skipped! Triton not found")
     @skipCPUIf(not HAS_CPU, "Skipped! Supported CPU compiler not found")
+    @unittest.skipIf(TEST_WITH_ASAN, "Skipped under ASAN")
```
Any reason for this skipping?

eellison replied:
I think it's failing now because we run constant folding on empty kernels, which breaks ASAN. If you look around there are a number of ASAN skips. I don't think it's providing a lot of value here..

eellison committed:
Update on "Add A Pass to Fold Tensors With a Uniform Value, match sdpa on a few models"



Adds a Constant Folding pass to the joint graph only targeting tensors which can be replaced with a single value, and then removes no-ops from the graph. This allows us to match sdpa in BertForMaskedLM, AlbertForMaskedLM, and LayoutLMForMaskedLM.

BertForMaskedLM 
Perf: 1.6853 -> 1.933, Memory: 0.9462 -> 1.41

AlbertForMaskedLM
Perf: 1.6620 -> 1.761, Memory: 1.257 -> 1.94

LayoutLMForMaskedLM
Perf: (non cudagraphs) 1.6991 -> 1.939x, Memory: 0.9624 -> 1.50

MobileBertForMaskedLM
Perf: 1.864x -> 1.941x, Memory: 0.94 -> 1.03



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

eellison committed:
Update on "Add A Pass to Fold Tensors With a Uniform Value, match sdpa on a few models"



Adds a Constant Folding pass to the joint graph only targeting tensors which can be replaced with a single value, and then removes no-ops from the graph. This allows us to match sdpa in BertForMaskedLM, AlbertForMaskedLM, and LayoutLMForMaskedLM.

BertForMaskedLM 
Perf: 1.6853 -> 1.933, Memory: 0.9462 -> 1.41

AlbertForMaskedLM
Perf: 1.6620 -> 1.761, Memory: 1.257 -> 1.94

LayoutLMForMaskedLM
Perf: (non cudagraphs) 1.6991 -> 1.939x, Memory: 0.9624 -> 1.50

MobileBertForMaskedLM
Perf: 1.864x -> 1.941x, Memory: 0.94 -> 1.03



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

eellison committed:
Update on "Add A Pass to Fold Tensors With a Uniform Value, match sdpa on a few models"



Adds a Constant Folding pass to the joint graph only targeting tensors which can be replaced with a single value, and then removes no-ops from the graph. This allows us to match sdpa in BertForMaskedLM, AlbertForMaskedLM, and LayoutLMForMaskedLM.

BertForMaskedLM 
Perf: 1.6853 -> 1.933, Memory: 0.9462 -> 1.41

AlbertForMaskedLM
Perf: 1.6620 -> 1.761, Memory: 1.257 -> 1.94

LayoutLMForMaskedLM
Perf: (non cudagraphs) 1.6991 -> 1.939x, Memory: 0.9624 -> 1.50

MobileBertForMaskedLM
Perf: 1.864x -> 1.941x, Memory: 0.94 -> 1.03



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

eellison committed:
Update on "Add A Pass to Fold Tensors With a Uniform Value, match sdpa on a few models"



Adds a Constant Folding pass to the joint graph only targeting tensors which can be replaced with a single value, and then removes no-ops from the graph. This allows us to match sdpa in BertForMaskedLM, AlbertForMaskedLM, and LayoutLMForMaskedLM.

BertForMaskedLM 
Perf: 1.6853 -> 1.933, Memory: 0.9462 -> 1.41

AlbertForMaskedLM
Perf: 1.6620 -> 1.761, Memory: 1.257 -> 1.94

LayoutLMForMaskedLM
Perf: (non cudagraphs) 1.6991 -> 1.939x, Memory: 0.9624 -> 1.50

MobileBertForMaskedLM
Perf: 1.864x -> 1.941x, Memory: 0.94 -> 1.03



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

eellison committed:
Update on "Add A Pass to Fold Tensors With a Uniform Value, match sdpa on a few models"



Adds a Constant Folding pass to the joint graph only targeting tensors which can be replaced with a single value, and then removes no-ops from the graph. This allows us to match sdpa in BertForMaskedLM, AlbertForMaskedLM, and LayoutLMForMaskedLM.

BertForMaskedLM 
Perf: 1.6853 -> 1.933, Memory: 0.9462 -> 1.41

AlbertForMaskedLM
Perf: 1.6620 -> 1.761, Memory: 1.257 -> 1.94

LayoutLMForMaskedLM
Perf: (non cudagraphs) 1.6991 -> 1.939x, Memory: 0.9624 -> 1.50

MobileBertForMaskedLM
Perf: 1.864x -> 1.941x, Memory: 0.94 -> 1.03



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

eellison committed:
Update on "Add A Pass to Fold Tensors With a Uniform Value, match sdpa on a few models"



Adds a Constant Folding pass to the joint graph only targeting tensors which can be replaced with a single value, and then removes no-ops from the graph. This allows us to match sdpa in BertForMaskedLM, AlbertForMaskedLM, and LayoutLMForMaskedLM.

BertForMaskedLM 
Perf: 1.6853 -> 1.933, Memory: 0.9462 -> 1.41

AlbertForMaskedLM
Perf: 1.6620 -> 1.761, Memory: 1.257 -> 1.94

LayoutLMForMaskedLM
Perf: (non cudagraphs) 1.6991 -> 1.939x, Memory: 0.9624 -> 1.50

MobileBertForMaskedLM
Perf: 1.864x -> 1.941x, Memory: 0.94 -> 1.03



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

eellison committed:
Update on "Add A Pass to Fold Tensors With a Uniform Value, match sdpa on a few models"



Adds a Constant Folding pass to the joint graph only targeting tensors which can be replaced with a single value, and then removes no-ops from the graph. This allows us to match sdpa in BertForMaskedLM, AlbertForMaskedLM, and LayoutLMForMaskedLM.

BertForMaskedLM 
Perf: 1.6853 -> 1.933, Memory: 0.9462 -> 1.41

AlbertForMaskedLM
Perf: 1.6620 -> 1.761, Memory: 1.257 -> 1.94

LayoutLMForMaskedLM
Perf: (non cudagraphs) 1.6991 -> 1.939x, Memory: 0.9624 -> 1.50

MobileBertForMaskedLM
Perf: 1.864x -> 1.941x, Memory: 0.94 -> 1.03



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

eellison committed:
Update on "Add A Pass to Fold Tensors With a Uniform Value, match sdpa on a few models"



Adds a Constant Folding pass to the joint graph only targeting tensors which can be replaced with a single value, and then removes no-ops from the graph. This allows us to match sdpa in BertForMaskedLM, AlbertForMaskedLM, and LayoutLMForMaskedLM.

BertForMaskedLM 
Perf: 1.6853 -> 1.933, Memory: 0.9462 -> 1.41

AlbertForMaskedLM
Perf: 1.6620 -> 1.761, Memory: 1.257 -> 1.94

LayoutLMForMaskedLM
Perf: (non cudagraphs) 1.6991 -> 1.939x, Memory: 0.9624 -> 1.50

MobileBertForMaskedLM
Perf: 1.864x -> 1.941x, Memory: 0.94 -> 1.03



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

eellison commented:
@pytorchbot merge
